### ISSUE TITLE
Fedora zfs-testing 2.0.0-rc6 - zfs-dkms and zfs-dracut rpms do not appear to update kernel, initramfs

### ISSUE THREAD

**Issue by averyfreeman** [openzfs/zfs#11210]
### System information

Type | Version/Name
Linux | Fedora release 33 (Thirty Three)
Distribution Name	| Fedora
Distribution Version	| 33
Linux Kernel	|  5.8.18-300.fc33.x86_64
GNU/Linux
Architecture	| x86_64
ZFS Version	| 2.0.0-rc6
SPL Version	| 2.0.0-rc6

#### Expected behavior: 
`zfs-dkms` package to compile kernel with appropriate kernel modules for zfs
`zfs-dracut` package to update initramfs with new dkms kernels for booting the system
#### Actual behavior: 
`zfs-dkms` and `zfs-dracut` packages do not appear to build kernel modules and initramfs for zfs.  Rebooting after installing packages results in non-booting system. 
### Describe how to reproduce the problem
compiled master (not sure if relevant, but including for completeness)
installed the following self-compiled rpms:
```
kmod-zfs-5.8.15-301.fc33.x86_64-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
kmod-zfs-5.8.15-301.fc33.x86_64-debuginfo-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
libnvpair1-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
libuutil1-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
libzfs2-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
libzfs2-devel-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
libzpool2-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
python3-pyzfs-2.0.0-rc1_212_g4352edaaf.fc33.noarch.rpm
zfs-2.0.0-rc1_212_g4352edaaf.fc33.x86_64.rpm
zfs-dkms-2.0.0-rc1_212_g4352edaaf.fc33.noarch.rpm
zfs-dracut-2.0.0-rc1_212_g4352edaaf.fc33.noarch.rpm
```
opted to go with `zfs-testing` repo instead:
```
dnf install http://download.zfsonlinux.org/fedora/zfs-release.fc33.noarch.rpm
vim /etc/yum.repos.d/zfs.repo # changed zfs enabled=0 zfs-testing=1
dnf update  # this replaced self-compiled packages
ZFS on Linux for Fedora 33 - Testing                               275 kB/s | 101 kB     00:00    
Dependencies resolved.
===================================================================================================
 Package                 Architecture    Version                        Repository            Size
===================================================================================================
Upgrading:
 kitty                   x86_64          0.19.2-1.fc33                  updates              3.6 M
 kitty-terminfo          noarch          0.19.2-1.fc33                  updates               22 k
 libnfsidmap             x86_64          1:2.5.2-1.rc1.fc33             updates               62 k
 libnvpair1              x86_64          2.0.0-rc6.fc33                 zfs-testing           40 k
 libuutil1               x86_64          2.0.0-rc6.fc33                 zfs-testing           33 k
 libzfs2                 x86_64          2.0.0-rc6.fc33                 zfs-testing          236 k
 libzfs2-devel           x86_64          2.0.0-rc6.fc33                 zfs-testing          342 k
 libzpool2               x86_64          2.0.0-rc6.fc33                 zfs-testing          1.3 M
 nfs-utils               x86_64          1:2.5.2-1.rc1.fc33             updates              418 k
 python3-pyzfs           noarch          2.0.0-rc6.fc33                 zfs-testing          117 k
 zfs                     x86_64          2.0.0-rc6.fc33                 zfs-testing          624 k
 zfs-dkms                noarch          2.0.0-rc6.fc33                 zfs-testing          8.5 M
 zfs-dracut              noarch          2.0.0-rc6.fc33                 zfs-testing           21 k

Transaction Summary
===================================================================================================
Upgrade  13 Packages

Total download size: 15 M
```
removed kmod packages I had compiled: 
```
% rpm -qa | egrep 'kmod|kabi'
kmod-libs-27-3.fc33.x86_64
kmod-27-3.fc33.x86_64
kmod-zfs-5.8.15-301.fc33.x86_64-2.0.0-rc1_212_g4352edaaf.fc33.x86_64
kmod-zfs-5.8.15-301.fc33.x86_64-debuginfo-2.0.0-rc1_212_g4352edaaf.fc33.x86_64

% sudo rpm -e kmod-zfs-5.8.15-301.fc33.x86_64-2.0.0-rc1_212_g4352edaaf.fc33.x86_64 kmod-zfs-5.8.15-301.fc33.x86_64-debuginfo-2.0.0-rc1_212_g4352edaaf.fc33.x86_64
```
rebooted - non booting system - sorry I can't find proper journalctl to show error messages :(

I am guessing I was able to boot earlier because of zfs-kmod package compiled for 5.8.15-301.fc33.x86_64 kernel (?)

chrooted into non-booting system from live usb and ran this script (which I had previously placed at `/etc/kernel/postinst.d`, but removed thinking `zfs-dkms` and `zfs-dracut` packages would replace its functionality):
```
#!/usr/bin/bash
dkms build zfs/2.0.0
dkms install zfs/2.0.0
modprobe zfs
systemctl daemon-reload
dracut -fv --kver `uname -r`
/usr/sbin/grub2-mkconfig -o /boot/efi/EFI/fedora/grub.cfg
touch /.autolabel
```
Rebooted - system now boots -- both kernels I have installed:
```
 rpm -qa | grep kernel
kernel-modules-5.8.15-301.fc33.x86_64
kernel-modules-extra-5.8.15-301.fc33.x86_64
kernel-srpm-macros-1.0-3.fc33.noarch
kernel-devel-5.8.15-301.fc33.x86_64
libreport-plugin-kerneloops-2.14.0-13.fc33.x86_64
abrt-addon-kerneloops-2.14.4-7.fc33.x86_64
kernel-headers-5.8.18-300.fc33.x86_64
kernel-modules-5.8.18-300.fc33.x86_64
kernel-modules-extra-5.8.18-300.fc33.x86_64
kernel-devel-5.8.18-300.fc33.x86_64
kernel-core-5.8.15-301.fc33.x86_64
kernel-core-5.8.18-300.fc33.x86_64
kernel-5.8.18-300.fc33.x86_64
kernel-5.8.15-301.fc33.x86_64
```
Note:  I have disabled selinux, but have included `touch /.autorelabel` because part of the boot errors included failed audit messages

### Include any warning/errors/backtraces from the system logs
```
Sorry I can't find any non-booting dmesg / journalctl - I don't think my system is configured to save them ATM
```

**Comment by dioni21** [comment]
@averyfreeman did you have a kernel upgrade pending reboot at the same time you have installed/upgraded zfs?

**Comment by L117** [comment]
I have similar issue on Debian. There was pending kernel update (After kernel installation, dpkg asked me to reboot the system, I refused). Today I did another update, new zfs packages were installed, so I decided to finally make a reboot. To my surprise system became unbootable (Root on ZFS). After booting from live USB, I've chrooted into Debian on ZFS and tried to recreate initrd by executing `dpkg-reconfigure zfs-dkms` and `dpkg-reconfigure dracut`. First seem to work, when I drop to initrd shell of new kernel, I can see ZFS module loaded and it reports correct version (2.0.1). Second - does not, it prints `update-initramfs: deferring update (trigger activated)`, then builds image nevertheless, bui it have no ZFS-related systemd services and no `zpool`/`zfs` binaries. When I execute dracut directly, like this `dracut /boot/custom-initrd.img --force-add zfs -f 5.10.0-1`, it says there's no `zfs` dracut module. What is not true, `/usr/lib/dracut/modules.d/90zfs` directory exists and is not empty.

@dioni21 Any suggestions?

**Comment by dioni21** [comment]
@L117 I have no automatic solution yet.  What I do is take a lot of care before rebooting if I made any upgrade.

Sometimes a simple `dkms status`  helps.  Just make sure you have a compiled ZFS for the next (newest?) kernel.

In Fedora world, the right kernel is indeed compiled during `dkms.service`, but it is usually too late for me, since Window Manager starts before, and I have auto-login in my desktop.  Yes, I could add a manual dependency, but too lazy for that.  ;-)

**Comment by L117** [comment]
@dioni21 I have tried everything possible, then gave up and switched back to Gentoo (I already had dataset with it). 

Surprisingly, both ZFS and dracut work here exactly as expected, so it is clearly Debian's fault and not ZFS's one.

**Comment by averyfreeman** [comment]
> @averyfreeman did you have a kernel upgrade pending reboot at the same time you have installed/upgraded zfs?

Thanks for getting back to me - As far as I remember, no, I don't believe I did, but that was quite a while ago.

Shortly after sending that bug report, I switched to Ubuntu since it saves separate linux headers packages for each kernel release, making it far friendlier for ZFS than RHEL distros.  

Now I just build my own ZFS built-in kernels so I don't have to deal with missing modules.
